src/test/java/io/confluent/connect/elasticsearch/ElasticsearchWriterTest.java

```diff
@@ -260,24 +261,89 @@ public void testStringKeyedMap() throws Exception {
-    SinkRecord sinkRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, key, mapSchema, map, 0);
+    SinkRecord sinkRecord = new SinkRecord(TOPIC, PARTITION, null, key, null, map, 0);
 
     ElasticsearchWriter writer = initWriter(client, ignoreKey, ignoreSchema, routingFieldName);
     writeDataAndRefresh(writer, Collections.singletonList(sinkRecord));
 
     Collection<?> expectedRecords = Collections.singletonList(new ObjectMapper().writeValueAsString(map));
-    verifySearchResults(expectedRecords, TOPIC, routingFieldValue, ignoreKey, ignoreSchema);
+    verifySearchResults(expectedRecords, TOPIC, routingFieldValue.toString(), ignoreKey, ignoreSchema);
+  }
+
+  @Test
+  public void testIndexThenUpdate() throws Exception {
+    boolean ignoreKey = false;
+    boolean ignoreSchema = true;
+
+    String routingFieldName = "organizationId";
+    Long routingFieldValue = 1L;
+
+    Map<String, Object> contact = new HashMap<>();
+    contact.put(routingFieldName, routingFieldValue);
+    contact.put("firstName", "Echo");
+    contact.put("lastName", "Xu");
+
+    SinkRecord sinkRecordForIndex = new SinkRecord(TOPIC, PARTITION, null, key, null, contact, 0);
+
+    Map<String, Object> contactUpdate = new HashMap<>();
+    contactUpdate.put(routingFieldName, routingFieldValue);
+
+    // for routing
+    contactUpdate.put(routingFieldName, routingFieldValue);
+
+    Map<String, String> doc = new HashMap<>();
+    doc.put("firstName", "Echo-updated");
+
+    contactUpdate.put("doc", doc);
+
+    SinkRecord sinkRecordForUpdate = new SinkRecord(TOPIC, PARTITION, null, key, null, contactUpdate, 0);
+
+    ElasticsearchWriter writer = initWriter(client, ignoreKey, ignoreSchema, routingFieldName);
+    writeDataAndRefresh(writer, Arrays.asList(sinkRecordForIndex, sinkRecordForUpdate));
+
+    contact.put("firstName", "Echo-updated");
+
+    Collection<?> expectedRecords = Collections.singletonList(new ObjectMapper().writeValueAsString(contact));
+    verifySearchResults(expectedRecords, TOPIC, routingFieldValue.toString(), ignoreKey, ignoreSchema);
+  }
+
+  @Test
+  public void testUpsert() throws Exception {
+    boolean ignoreKey = false;
+    boolean ignoreSchema = true;
+
+    String routingFieldName = "organizationId";
+    Long routingFieldValue = 1L;
+
+    Map<String, Object> map = new HashMap<>();
+    map.put(routingFieldName, routingFieldValue);
+
+    // for routing
+    map.put(routingFieldName, routingFieldValue);
+    map.put("doc_as_upsert", true);
+
+
+    Map<String, String> doc = new HashMap<>();
+    doc.put("firstName", "Echo-updated");
+
+    map.put("doc", doc);
+
+    SinkRecord sinkRecord = new SinkRecord(TOPIC, PARTITION, null, key, null, map, 0);
+
+    ElasticsearchWriter writer = initWriter(client, ignoreKey, ignoreSchema, routingFieldName);
+    writeDataAndRefresh(writer, Collections.singletonList(sinkRecord));
+
+    Collection<?> expectedRecords = Collections.singletonList("{\"firstName\":\"Echo-updated\"}");
```

Comment:
Any reason for not using an ObjectMapper to write out the json here as in L315?

Reply:
Just do not want to create a new object and use ObjectWrapper to make it a string, in this case, the expected output is not exactly same as the `map ` object.